Fixes part of #4272: Added Tests for EventLogSubject.

theayushyadav11 opened the pull request:
Fixes part of #4272: 

This pull request introduces new tests for the `EventLogSubject` class in the `Oppia` Android project. The main changes include the addition of the `EventLogSubjectTest` class and the corresponding build configuration updates.

### New Tests for `EventLogSubject`:

* [`testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt`](diffhunk://#diff-346c08638dcbe5a2f3c1911f47813e71dcd5de2c528f19bdb326555aa9767712R1-R298): Added a comprehensive set of tests for the `EventLogSubject` class, covering various scenarios such as timestamp matching, priority checks, profile ID validation, and language selection verification.

### Build Configuration Updates:

* [`testing/src/test/java/org/oppia/android/testing/logging/BUILD.bazel`](diffhunk://#diff-c0153ba759e957520c4c5d5eb6221aab875f577074f14fe0f670cda8761a8f34R60-R79): Added a new build target for `EventLogSubjectTest`, including necessary dependencies for running the tests.

### Code Cleanup:

* [`testing/src/main/java/org/oppia/android/testing/logging/EventLogSubject.kt`](diffhunk://#diff-38666dc35e58390379b0604e36be3ee798c70f4d4d2070623a602a6023fec12aL73-L74): Removed outdated TODO comments related to adding tests for the `EventLogSubject` class.

All tests has been added succesfully.

#### Screenshot of all the passing tests:
![image](https://github.com/user-attachments/assets/1414a9c3-165d-4b5c-ae2e-7213406f99bb)


## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Commit:
Added some tests for EventLogSubject

Commit:
Removed Todo

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 commented:
Hi @adhiamboperes PTAL, as mentioned in the PR description these tests are not complete , once reviewed I would further add the remaining tests as `EventLogSubject` has the extensive scope.

Commit:
reformatting

Commit:
Merge remote-tracking branch 'origin/eventlog-subject-Test' into eventlog-subject-Test

Commit:
removed buildifier

adhiamboperes commented on testing/src/main/java/org/oppia/android/testing/logging/EventLogSubject.kt:
Please remove this file from `test_file_exemptions.textproto` so code coverage analysis can run.

adhiamboperes commented on testing/src/main/java/org/oppia/android/testing/logging/EventLogSubject.kt:
```diff
@@ -70,8 +70,6 @@ import org.oppia.android.app.model.UserTypeAnswer
 import org.oppia.android.app.model.WrittenTranslationLanguageSelection
 import org.oppia.android.testing.logging.EventLogSubject.Companion.assertThat
 
-// TODO(#4272): Add tests for this class.
```
Please do a global search and remove any other occurrances of this `TODO(#4272)`.

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,298 @@
+@RunWith(JUnit4::class)
+class EventLogSubjectTest {
+  @Test
+  fun testHasTimeStamp_withTimeStamp_matchesTimeStamp() {
```
All the tests in this class needs to be renamed to follow our preferred format: https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#guidelines-for-testing, e.g. `testEventLogSubject_matchesCorrectTimestamp`
`testEventLogSubject_failsOnUnmatchingTimestamp`

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,298 @@
+      .isEqualTo(123456789)
+  }
+
+  @Test(expected = AssertionError::class)
+  fun testHasTimeStamp_withDifferentTimeStamp_fails() {
+    val eventLog = EventLog.newBuilder()
+      .setTimestamp(123456789)
+      .build()
+
+    EventLogSubject.assertThat(eventLog)
+      .hasTimestampThat()
+      .isEqualTo(987654321)
+  }
```
We don't use this pattern when testing. Prefer something likee:
```suggestion
  @Test
  fun testEventLogSubject_failsOnUnmatchingTimestamp() {
    val eventLog = EventLog.newBuilder()
      .setTimestamp(123456789)
      .build()

 assertThrows<AssertionError>() {
      EventLogSubject.assertThat(eventLog)
      .hasTimestampThat()
      .isEqualTo(987654321)
    }   
  }
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,298 @@
+  }
+
+  @Test(expected = AssertionError::class)
+  fun testIsEssentialPriority_withDifferentPriority_fails() {
```
```suggestion
  fun testEventLogSubject_matchEssentialPriorityWithDifferentPriority_fails() {
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,298 @@
+  }
+
+  @Test
+  fun testHasNoProfileId_withNoProfileId() {
```
```suggestion
  fun testEventLogSubject_eventWithNoProfileId_returnsNoProfileId() {
```
Ditto test below

adhiamboperes reviewed (changes requested):
Thanks @theayushyadav11!

My main suggestion is to change how the tests are named to reflect `classUnderTest_condition1BeingTested_condition2ifApplicable_expectedOutcome()`

Please rename these tests before adding more so that you can get a good grasp of the pattern.

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/BUILD.bazel:
```diff
@@ -57,3 +57,21 @@ oppia_android_test(
+    test_class = "org.oppia.android.testing.logging.EventLogSubjectTest",
+    test_manifest = "//testing:test_manifest",
+    deps = [
+        "//:dagger",
+        "//testing",
```
Are these two deps really required here?

theayushyadav11 commented:
Thanks @adhiamboperes for the review, I'll implement the changes as required.

Commit:
Resolved comments

theayushyadav11 commented:
Hi @adhiamboperes, PTAL! Let me know if I should continue writing tests for other methods or if you have any other suggestions. Thanks!

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -123,14 +130,15 @@ class EventLogSubjectTest {
 
   @Test
-  fun testHasAppLanguageSelectionThat_withAppLanguageSelection_matchesAppLanguageSelection() {
+  fun testEventLogSubject_hasSameAppLanguageSelection() {
```
```suggestion
  fun testEventLogSubject_matchesAppLanguageSelection() {
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -216,14 +226,15 @@ class EventLogSubjectTest {
 
   @Test
-  fun testHasOpenExplorationActivityContext_withMatchingContext_passes() {
+  fun testEventLogSubject_hasOpenExplorationActivityContextPresent() {
```
For this and below, remove redundant verb.
```suggestion
  fun testEventLogSubject_hasOpenExplorationActivityContext() {
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -235,17 +246,18 @@ class EventLogSubjectTest {
-  @Test(expected = AssertionError::class)
-  fun testHasOpenExplorationActivityContext_withDifferentContext_fails() {
+  @Test
+  fun testEventLogSubject_openExplorationActivityContextAbsent_fails() {
```
For this and below, rename:
```suggestion
  fun testEventLogSubject_missingExplorationActivityContext_fails() {
```

adhiamboperes reviewed (changes requested):
@theayushyadav11, I have left a few corrections but please feel free to continue adding tests.

Commit:
renamed functions

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

Commit:
added more tests

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

Commit:
added more tests

Commit:
added more tests

Commit:
Added all tests

theayushyadav11 commented:
Hi @adhiamboperes, I have added all the tests for the `EventLogSubject`  PTAL  and please tell me if any more changes are needed. Thanks!

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 commented:
Hi @adhiamboperes , this PR is ready for review PTAL and please suggest the needed changes in the code.
Thanks!

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 commented:
Hi @adhiamboperes , this PR is ready for review PTAL and please suggest the needed changes in the code.
Thanks!

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_matchesPriorityEssential() {
```
```suggestion
  fun testEventLogSubject_withPriorityEssential_passes() {
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_matchesPriorityOptional() {
```
```suggestion
  fun testEventLogSubject_withPriorityOptional_passes() {
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_matchesProfileIdPresent() {
```
```suggestion
  fun testHasProfileIdThat_eventWithProfileId_returnsProfileIdSubject {
```

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_matchesAppLanguageSelection() {
```
```suggestion
  fun testEventLogSubject_eventWithAppLanguageSelection_returnsAppLanguageSelectionSubject() {
```

theayushyadav11 replied:
It crosses the word limit per line.

adhiamboperes replied:
Please see the edited suggestion.

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_matchesWrittenTranslationLanguageSelection() {
```
```suggestion
  fun testHasWrittenTranslationLanguageSelectionThat_eventWithWrittenTranslationLanguageSelection_returnsWrittenTranslationLanguageSelectionSubject() {
```

theayushyadav11 replied:
Ditto

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_matchesAudioTranslationLanguageSelection() {
```
```suggestion
  fun testHasAudioTranslationLanguageSelectionThat_eventWithAudioTranslationLanguageSelection_returnsAudioTranslationLanguageSelectionSubject() {
```

theayushyadav11 replied:
Ditto

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_hasOpenExplorationActivityContext() {
```
```suggestion
  fun testEventLogSubject_hasOpenExplorationActivityContext_passes() {
```

adhiamboperes replied:
Add tests for:

1. `testHasOpenExplorationActivityContextThat_contextIsExploration_returnsExplorationContextSubject`
2. `testHasOpenExplorationActivityContextThat_blockIsProvided_executesWithCorrectSubject`

adhiamboperes commented on testing/src/test/java/org/oppia/android/testing/logging/EventLogSubjectTest.kt:
```diff
@@ -0,0 +1,3939 @@
+  }
+
+  @Test
+  fun testEventLogSubject_hasOpenInfoTabContext() {
```
Like with the exploration context test above, we should test for 3 outcomes, in addition to the failure case:
1. `_whenContextIsInfoTab_passes` ==  EventLogSubject.assertThat(eventLog)
      .hasOpenInfoTabContext()

2. `_returnsTopicContextSubject`  

```kotlin
 val eventLog = EventLog.newBuilder()
    .setContext(
      EventLog.Context.newBuilder()
        .setOpenInfoTab(
          TopicContext.newBuilder().setTopicId("info_tab_topic")
        )
    )
    .build()

  val subject = EventLogSubject.assertThat(eventLog)
    .hasOpenInfoTabContextThat()

subject.hasTopicIdThat().isEqualTo("info_tab_topic")
```

3. `_executesBlockWithCorrectSubject`

  ```kotlin
val eventLog = EventLog.newBuilder()
    .setContext(
      EventLog.Context.newBuilder()
        .setOpenInfoTab(TopicContext.newBuilder().setTopicId("test_topic"))
    )
    .build()

  EventLogSubject.assertThat(eventLog)
    .hasOpenInfoTabContextThat {
      hasTopicIdThat().isEqualTo("test_topic")
    }
```

This is because every contxt subject has these 3 functions, and this will increase the code coverage of the class. Ditto to the remaining contexts in the class.

adhiamboperes reviewed (changes requested):
Thanks @theayushyadav11, for setting up tests for this extremely large class! 

Primarily, the test setup is good, and the coverage is great. I have some naming pattern suggestions and I have identified some missing test coverage for contexts. To clarify:

Contexts tyocally have 3 subject functions, `fun hasSomeContext()`, `fun hasSomeContextThat()`, and `hasSomeContextThat(block: TopicContextSubject.() -> Unit)`.

This PR adequatel covers the `fun hasSomeContext()`, functions, although I would prefer them renamed to `testEventLogSubject_hasSomeContext_passes`, to follow the `functionUnderTest_condition_expectedBehavior` pattern.

As for the 2 that are not covered, I am fine with those going in a follow up PR, to avoid adding a significant amount of work to this PR. This means changing the PR title to "Fix Part of". 

I am happy to merge this PR with only the test name updates, since it passes the required code coverage threshold.

theayushyadav11 committed:
Merge branch 'develop' into eventlog-subject-Test

theayushyadav11 commented:
Hi  @adhiamboperes, I only have to change the test names in this pr as suggested (some are exceeding 100 letters) to make this pr merged and will raise a new pr for the other two types of functions which are not covered in coverage and will be looking at the `html` file and add the tests for those...please clarify so that I wrap up this asap.

adhiamboperes commented:
@theayushyadav11,  tecgnically, yes. You need to address all comments except the ones for adding new tests for `hasSomeContextThat(),` and `hasSomeContextThat(block: TopicContextSubject.() -> Unit)`. You can edit the long tests names as per my latest suggestion.

>  will be looking at the `html` file and add the tests for those...please clarify so that I wrap up this asap.

Which **html** file are you referring to?

theayushyadav11 commented:
> Which **html** file are you referring to?

The file which generates after running code coverage locally.

Commit:
renamed tests

Commit:
Merge remote-tracking branch 'origin/eventlog-subject-Test' into eventlog-subject-Test

theayushyadav11 commented:
PTAL @adhiamboperes.

adhiamboperes reviewed (approved):
Thanks @theayushyadav11! This LGTM.

adhiamboperes commented:
Would like add a reminder: Always respond to reviewrs comments, even if it is with, "Done.". This lets us know that you have addressed the comment, and we can verify and resolve the comment.

theayushyadav11 commented:
> Would like add a reminder: Always respond to reviewrs comments, even if it is with, "Done.". This lets us know that you have addressed the comment, and we can verify and resolve the comment.


Apologies! I’ll make sure to take care of this from next time, as it will definitely make the review process easier.
